File: src/lib/agentConnect.ts
import { invoke } from "@tauri-apps/api/core";

// "Connect agents": registers the bundled status hooks for Claude Code, Codex, and
// OpenCode so their panes report a Task line and a true Running/Waiting/Idle badge.

export const CONNECT_AGENTS_EVENT = "termfleet:connect-agents";

export type AgentConnectState = "connected" | "not-connected" | "absent" | "blocked" | "error";

export interface AgentConnectSummary {
  ok: boolean;
  connected: number;
  needed: number;
  results: Array<{ agent: string; state: AgentConnectState; detail: string }>;
}

/** Agents that are installed for this user but not connected to TermFleet yet. */
export function agentsNeedingConnection(summary: AgentConnectSummary | null) {
  return (summary?.results ?? [])
    .filter((item) => item.state === "not-connected")
    .map((item) => item.agent);
}

export function agentConnectResultText(summary: AgentConnectSummary) {
  const names: Record<string, string> = { claude: "Claude Code", codex: "Codex", opencode: "OpenCode" };
  const connected = summary.results
    .filter((item) => item.state === "connected")
    .map((item) => names[item.agent] ?? item.agent);
  const problems = summary.results.filter((item) => item.state === "blocked" || item.state === "error");
  if (problems.length) return problems.map((item) => item.detail).join(" · ");
  if (!connected.length) return "No Claude Code, Codex, or OpenCode setup found for this user";
  return `Connected: ${connected.join(", ")}. New agent sessions will report their status.`;
}

export async function runAgentConnect(check: boolean): Promise<AgentConnectSummary> {
  const raw = await invoke<string>("agents_connect", { check });
  return JSON.parse(raw) as AgentConnectSummary;
}

File: src-tauri/src/agent_connect.rs
//! "Connect agents": registers the bundled Claude Code / Codex / OpenCode status
//! hooks for this user, so agent panes report a real Task line and a true
//! Running / Waiting / Idle badge in a downloaded build, not only in a source
//! checkout. The work lives in `termfleet-connect-agents.mjs` (shipped as a
//! bundle resource); this only locates it and runs it with the user's Node.js.

use std::path::PathBuf;
use std::process::{Command, Stdio};
use tauri::Manager;

const SCRIPT: &str = "agent-hooks/termfleet-connect-agents.mjs";

fn connect_script(app: &tauri::AppHandle) -> Option<PathBuf> {
    if let Ok(path) = app
        .path()
        .resolve(SCRIPT, tauri::path::BaseDirectory::Resource)
    {
        if path.exists() {
            return Some(path);
        }
    }
    // Development builds run straight from the checkout.
    let dev = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../scripts/termfleet-connect-agents.mjs");
    dev.exists().then_some(dev)
}

/// Runs the connector (`check` = report only) and returns its JSON summary.
#[tauri::command]
pub async fn agents_connect(app: tauri::AppHandle, check: bool) -> Result<String, String> {
    let script = connect_script(&app)
        .ok_or_else(|| "The agent connector is missing from this build.".to_string())?;
    tauri::async_runtime::spawn_blocking(move || {
        let mut command = Command::new("node");
        command.arg(&script).arg("--json");
        if check {
            command.arg("--check");
        }
        let output = command
            .stdin(Stdio::null())
            .output()
            .map_err(|_| "Connecting agents needs Node.js 20 or newer on your PATH.".to_string())?;
        let stdout = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if stdout.starts_with('{') {
            Ok(stdout)
        } else {
            Err(format!(
                "The agent connector failed: {}",
                String::from_utf8_lossy(&output.stderr).trim()
            ))
        }
    })
    .await
    .map_err(|error| error.to_string())?
}
